java/ql/src/Security/CWE/CWE-036/OpenStream.ql

```diff
@@ -0,0 +1,58 @@
+/**
+ * @name Calling openStream on URLs created from remote source can lead to file disclosure
+ * @description If openStream is called on a java.net.URL, that was created from a remote source
+ *              an attacker can try to pass absolute URLs starting with file:// or jar:// to access
+ *              local resources in addition to remote ones.
+ * @kind path-problem
+ */
+
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.FlowSources
+import DataFlow::PathGraph
+
+class URLConstructor extends ClassInstanceExpr {
+  URLConstructor() { this.getConstructor().getDeclaringType() instanceof TypeUrl }
+
+  Expr stringArg() {
+    // Query only in URL's that were constructed by calling the single parameter string constructor.
+    if
+      this.getConstructor().getNumberOfParameters() = 1 and
+      this.getConstructor().getParameter(0).getType().getName() = "String"
+    then result = this.getArgument(0)
+    else none()
```

Comment:
The pattern `if A then B else none()` is equivalent to `A and B`.  Also, instead of `T.getName() = "String"` you can write `T instanceof TypeString`, which is a bit nicer.

Reply:
Thanks for the feedback!